praw/exceptions.py

```diff
@@ -1,38 +1,96 @@
 """PRAW exception classes.
 
-Includes two main exceptions: :class:`.APIException` for when something goes
-wrong on the server side, and :class:`.ClientException` when something goes
-wrong on the client side. Both of these classes extend :class:`.PRAWException`.
+Includes two main exceptions: :class:`.RedditAPIException` for when something
+goes wrong on the server side, and :class:`.ClientException` when something
+goes wrong on the client side. Both of these classes extend
+:class:`.PRAWException`.
 
 All other exceptions are subclassed from :class:`.ClientException`.
 
 """
-from typing import Optional
+from typing import List, Optional, TypeVar, Union
+
+_RedditErrorItem = TypeVar("_RedditErrorItem")
 
 
 class PRAWException(Exception):
     """The base PRAW Exception that all other exception classes extend."""
 
 
-class APIException(PRAWException):
-    """Indicate exception that involve responses from Reddit's API."""
+class RedditErrorItem:
+    """Represents a single error returned from Reddit's API."""
+
+    @property
+    def error_message(self):
+        """Get the completed error message string."""
+        error_str = "{}: '{}'".format(self.error_type, self.message)
+        if self.field:
+            error_str += " on field '{}'".format(self.field)
+        return error_str
 
-    def __init__(self, error_type: str, message: str, field: Optional[str]):
-        """Initialize an instance of APIException.
+    def __init__(
+        self, error_type: str, message: str, field: Optional[str] = None
+    ):
+        """Instantiate an error item.
 
         :param error_type: The error type set on Reddit's end.
         :param message: The associated message for the error.
-        :param field: The input field associated with the error if available.
+        :param field: The input field associated with the error, if available.
         """
-        error_str = "{}: '{}'".format(error_type, message)
-        if field:
-            error_str += " on field '{}'".format(field)
-
-        super().__init__(error_str)
         self.error_type = error_type
         self.message = message
         self.field = field
 
+    def __eq__(self, other: Union[_RedditErrorItem, List[str]]):
+        """Check for equality."""
+        if isinstance(other, RedditErrorItem):
+            return (self.error_type, self.message, self.field) == (
+                other.error_type,
+                other.message,
+                other.field,
+            )
+        return NotImplemented
```

Comment:
Consider calling `super` here to rely on pythons default behavior when they are of different instance types. I believe it's not expected for `==` comparison to ever raise an exception.

Reply:
This needs to return `NotImplemented` if you want it to be consistent with how 3.7’s dataclasses module works.

The `NotImplement` constant should not be confused with the `NotImplementedError` exception.

Reply:
Can you provide more context on the dataclasses module? I'm not sure what you mean.

Reply:
@bboe Here's an example: 
```python
import dataclasses
import inspect
from typing import Union

@dataclasses.dataclass(order=True)
class Person:
  name: str = dataclasses.field(compare=False)
  age: Union[int, float]
  relatives: dict = dataclasses.field(compare=False, default_factory=dict)

p1 = Person("Test", 25)
p2 = Person("Testing", 15)
print(p1>p2) # True
print(p1.__lt__(None)) # NotImplemented
print(p1<None) # TypeError('<' not supported between instances of 'Person' and 'NoneType')
```